tests/testsuite/doc.rs

```diff
@@ -1638,3 +1639,145 @@ fn crate_versions_flag_is_overridden() {
+    let dummy_project = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "foo"
+            version = "1.2.4"
+            authors = []
+        "#,
+        )
+        .file("src/lib.rs", "//! These are the docs!")
+        .build();
+
+    dummy_project.cargo("doc").run();
+
+    let fingerprint: RustDocFingerprint = serde_json::from_str(
+        dummy_project
+            .read_file(
+                std::path::PathBuf::from("target")
+                    .join("doc")
+                    .join(".rustdoc_fingerprint.json")
+                    .to_str()
+                    .expect("Malformed path"),
+            )
+            .as_str(),
+    )
+    .expect("JSON Serde fail");
+
+    // Check that the fingerprint contains the actual rustc version
+    // which has been used to compile the docs.
+    let output = std::process::Command::new("rustc")
+        .arg("-vV")
+        .stdout(std::process::Stdio::piped())
+        .output()
+        .expect("Failed to get actual rustc verbose version");
+    assert_eq!(
+        fingerprint.rustc_verbose_version,
+        (String::from_utf8_lossy(&output.stdout).as_ref())
+    );
+
+    // As the test shows above. Now we have generated the `doc/` folder and inside
+    // the rustdoc fingerprint file is located with the correct rustc version.
+    // So we will remove it and create a new fingerprint with an old rustc version
+    // inside it. We will also place a bogus file inside of the `doc/` folder to ensure
+    // it gets removed as we expect on the next doc compilation.
+    fs::remove_file(
+        dummy_project
+            .root()
+            .join("target")
+            .join("doc")
+            .join(".rustdoc_fingerprint.json"),
+    )
+    .expect("Failed to read fingerprint on tests");
+
+    fs::write(
+        dummy_project
+            .root()
+            .join("target")
+            .join("doc")
+            .join(".rustdoc_fingerprint.json"),
+        stable_rustc_verbose_version,
+    )
+    .expect("Error writing old rustc version");
+
+    fs::write(
+        dummy_project
+            .root()
+            .join("target")
+            .join("doc")
+            .join("bogus_file"),
+        String::from("This is a bogus file and should be removed!"),
+    )
+    .expect("Error writing test bogus file");
+
+    // Now if we trigger another compilation, since the fingerprint contains an old version
+    // of rustc, cargo should remove the entire `/doc` folder (including the fingerprint)
+    // and generating another one with the actual version.
+    // It should also remove the bogus file we created above
+    dummy_project.change_file("src/lib.rs", "//! These are the docs 2!");
+    dummy_project.cargo("doc").run();
+
+    assert!(fs::File::open(
+        dummy_project
+            .root()
+            .join("target")
+            .join("doc")
+            .join("bogus_file")
+    )
+    .is_err());
```

Comment:
This can be simplified:

```suggestion
    assert!(!dummy_project.build_dir().join("doc/bogus_file").exists());
```